bundles/org.openhab.binding.folderwatcher/src/main/java/org/openhab/binding/folderwatcher/internal/handler/FolderWatcherHandler.java

```diff
@@ -0,0 +1,291 @@
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.TimeUnit;
+
+import org.apache.commons.net.ftp.FTPClient;
+import org.apache.commons.net.ftp.FTPConnectionClosedException;
+import org.apache.commons.net.ftp.FTPFile;
+import org.apache.commons.net.ftp.FTPReply;
+import org.apache.commons.net.ftp.FTPSClient;
+import org.eclipse.jdt.annotation.NonNullByDefault;
+import org.eclipse.jdt.annotation.Nullable;
+import org.eclipse.smarthome.config.core.ConfigConstants;
+import org.eclipse.smarthome.core.thing.ChannelUID;
+import org.eclipse.smarthome.core.thing.Thing;
+import org.eclipse.smarthome.core.thing.ThingStatus;
+import org.eclipse.smarthome.core.thing.ThingStatusDetail;
+import org.eclipse.smarthome.core.thing.binding.BaseThingHandler;
+import org.eclipse.smarthome.core.types.Command;
+import org.openhab.binding.folderwatcher.internal.common.WatcherCommon;
+import org.openhab.binding.folderwatcher.internal.config.FolderWatcherConfiguration;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * The {@link FolderWatcherHandler} is responsible for handling commands, which are
+ * sent to one of the channels.
+ *
+ * @author Alexandr Salamatov - Initial contribution
+ */
+@NonNullByDefault
+public class FolderWatcherHandler extends BaseThingHandler {
+    private final Logger logger = LoggerFactory.getLogger(FolderWatcherHandler.class);
+    private @Nullable FolderWatcherConfiguration config;
+    private File currentFtpListingFile;
+    private @Nullable ScheduledFuture<?> executionJob, initJob;
+    private FTPClient ftp = new FTPClient();
+    private ArrayList<String> currentFtpListing = new ArrayList<String>();
+    private ArrayList<String> previousFtpListing = new ArrayList<String>();
+
+    public FolderWatcherHandler(Thing thing) {
+        super(thing);
+        currentFtpListingFile = new File(ConfigConstants.getUserDataFolder() + File.separator + "FolderWatcher"
+                + File.separator + thing.getUID().getAsString().replace(':', '_') + ".data");
+        try {
+            previousFtpListing = WatcherCommon.initStorage(currentFtpListingFile);
+        } catch (IOException e) {
+            logger.debug("Can't write file: {}", currentFtpListingFile, e);
+        }
+    }
+
+    @Override
+    public void handleCommand(ChannelUID channelUID, Command command) {
+        logger.debug("handleCommand for channel {}: {}", channelUID.getId(), command.toString());
+        logger.debug("Command sending not supported by this binding");
+    }
+
+    @Override
+    public void initialize() {
+        config = getConfigAs(FolderWatcherConfiguration.class);
+        updateStatus(ThingStatus.UNKNOWN);
+
+        if (config.pollInterval > 0) {
+            initJob = scheduler.scheduleWithFixedDelay(periodicConnectRunnable, 0, config.pollInterval,
+                    TimeUnit.SECONDS);
+            executionJob = scheduler.scheduleWithFixedDelay(periodicExecutionRunnable, config.pollInterval + 5,
```

Comment:
Better add that at the end of a successful connection attempt. This makes sure that the connection is established when the first refresh attempt is made